### PR TITLE
2.3 triadic refactoring

### PR DESCRIPTION
**Pull request by craigtaverner**
Triadic supported using combined TriadicSelectionPipe
- Merged TriadicBuild and TriadicPrope into TriadicSelection
- Fixed bugs around relationship type and direction
- Added much more extensive acceptance and integration tests
- Support positive predicates in addition to negative predicates

TriadicSelection as a single pipe makes it easier to share state between the build phase and the probe phase.

### CODE REVIEW

**Comment by craigtaverner** [comment]
Added the 'bug' label because this includes a fix for a case where we could choose to plan a triadic selection that was not in fact possible to plan. Specifically the relationship type in the pattern expression in the predicate is not tested against the relationship type in the original match expression. So 'MATCH (a)-[:KNOWS]->(b)-[:KNOWS]->(c) WHERE NOT (a)-->(c)' would be incorrectly planned with triadic.

**Comment by thobe** [comment]
Looks good.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/TriadicSelectionAcceptanceTest.scala [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    relate(nodes("b2"), nodes("b3"), relTypeBC)
+    relate(nodes("b3"), nodes("b4"), relTypeBC)
+    relate(nodes("b4"), nodes("b1"), relTypeBC)
+
+    nodes.toMap
+  }
+
+  // No predicate
+
+  test("should handle triadic friend of a friend") {
```
Comment (pontusmelke) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/TriadicSelectionAcceptanceTest.scala:76]: Should these tests assert that we actually use `TriadicSelection`? Something like  `result.executionPlanDescription().toString should include("TriadicSelection")` or the `use("TriadicSelection")` used in the integration test?

Comment (craigtaverner) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/TriadicSelectionAcceptanceTest.scala:76]: No. We have that assertion in the integration tests. These tests only assert that queries that are similar to triadic always return the correct answers, with both rule and cost planners (only only cost supported triadic). Also, as we expand the support for a wider range of queries, these test should be there to verify we don't break anything. So currently only a subset of them actually use triadic, many still do not. Some more could (with effort) and others never will.

**Pull request merged**